### ISSUE TITLE
DeepAID Interpreter for DeepLog

### ISSUE THREAD

**Issue by Fengrui-Liu** [dongtsi/DeepAID#2]
Hi, I read the fascinating paper, but the DeepLog demo confused me, in [timeseries_onehot.py](https://github.com/dongtsi/DeepAID/blob/main/deepaid/interpreters/timeseries_onehot.py) 
```
Loss = loss_accuracy + self.lbd * loss_determinism
optimizer.zero_grad()
Loss.backward()
....

print('Finish Interpretation after {} steps'.format(self.k), '(Final loss: %.2f,'%Loss.item(), "Time elasped: %.2fs)"%(end_time-begin_time)) 
return [IDX1, IDX2], REF
```

I notice that there is only once backward, by monitoring the Loss until interpreation finished, its value (Loss value) does not update and the optimizer seems not works right. The Deeplog demo runs in a different way from the original paper.

Have I missed anything? Thanks for your help!

**Comment by dongtsi** [comment]
Thanks for the wonderful question. 
The Implementation of Deeplog Interpreter is indeed a little bit different from what we described in Alg. 2 in our paper. 
Since Deeplog uses univariate time series which contains **discrete** values, it may not make sense to modify a value multiple times (**because you have to discretize it eventually**).
Therefore, for convenience, we only do the backpropagation once to obtain the gradient information of the time series X, and then use this gradient to "saliency test" and modify the dimension of X with the largest gradient.
In fact, in all the experiments of our paper, we fixed the number of modified dimensions (K) to 1, so we did not care too much about the "position of the loop" during multiple modifications 😄.
Of course, you could also modify the first dimension and then recalculate the gradient and then modify the second dimension (if K>1).
